### PR TITLE
feature: Display map that is updated with user's location

### PR DESCRIPTION
**Pull request by kyle-gonzales**
### Displaying the Map (finally)

I initially thought that I would have to do some crazy XML to Compose stuff with the AndroidView composable since a lot of the resources that I read stated that Google Maps doesn't have a Compose API yet. Thankfully, a Google Maps Compose Library was recently added last year.

- Added a google map composable to the main screen. It is displayed when driver mode is toggled on. (changes)
  - Swiping around the google map composable creates conflicts with the gesture controls of the menu drawer. To fix this, the menu drawer gestures are disabled when the menu drawer is closed. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/7a09227c179cbc8e692c5fa13242e72119090953#diff-e344f8c25f4e47d4996b0bcb53c80e9bd1ba8e91fa67886eb5e20d7e763659e5R540))
- Added light and dark mode map styles to the google map composable. However, the map styles are not yet fully customized. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/2aff7c856173c2eb905c9d021c64f8ac2201eaf4))
- Custom map UI settings such as the compass and the locationButton were added to the map composable. However, they don't seem to show on the screen. I think its an API problem tho. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/c66d28fb2c81ad2cd7021a116504dbfc72d54ae1#diff-e344f8c25f4e47d4996b0bcb53c80e9bd1ba8e91fa67886eb5e20d7e763659e5R360-R363)) 

### Getting Location Permissions

Without the location permissions, Google Maps cannot get updates of the user's location.

- Permissions are added to the manifest so we can  request for them in the app. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/a89ca68ca07dd2986e52f8f215795f685563cd61))
- Location permission requests are launched when the driving mode switch is toggled on. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/00ce84c9791c252e457df9032640089cbe302171#diff-e344f8c25f4e47d4996b0bcb53c80e9bd1ba8e91fa67886eb5e20d7e763659e5R140-R144))
  - I set up the permission requests in such a way that multiple permission requests can be launched when the driver mode is toggled on. This will be especially useful if we do plan to have a calling feature that requires microphone access or if we plan to continue tracking distance and time as a background service when the user closes the application. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/faa9c7f77d7bee816f945f24e6944012c238db29#diff-dc3bad78ace96db9e9c983fbadf2d801fa350087c6bc36e11e02759d5b57dc55R156-R169))
    - A permission is added to a permissionDialogQueue when the permission is denied. Each denied permission in the queue will launch an alert dialog. The permission is dequeued when the dialog closes.
  - If the permission requests have been granted, the driving mode can be toggled on.
  - If the user does not grant permissions, driving mode cannot be toggled on. However, we still allow the user to use other JeepNi features. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/bd8417c58f1fbb17cf20445ad04cccbecca4310c#diff-e344f8c25f4e47d4996b0bcb53c80e9bd1ba8e91fa67886eb5e20d7e763659e5R204-R224))
    - When the user denies the permission request once, an alert dialog is displayed indicating the importance of the permission request. If the request is launched and denied again, the permissions will be permanently denied. Whenever the user tries to toggle the driver mode on, the user will be prompted via an alert dialog to manually grant the permissions in the app's settings, which is launched when the user clicks the "Go to Settings" Button. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/bd8417c58f1fbb17cf20445ad04cccbecca4310c#diff-e344f8c25f4e47d4996b0bcb53c80e9bd1ba8e91fa67886eb5e20d7e763659e5R203-R224))
- The app data's [backup storage](https://stackoverflow.com/a/15744733/13907596) flag is changed to false. This allows the user's preferences and app configurations (especially, permissions) to be reset every time the application is reinstalled. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/3fbe550fd424013a7957ec05a900380e6048e68b#diff-7fa6aef292187a049f7a4d6060d8df3ba212d838789c78940bd363344b1c38cdR12))

### Getting User Location and Updating Map State

With the Map composable and Location permissions, user location can be obtained which can be used to update the user's marker on the map.

The [FusedLocationProvider](https://developers.google.com/android/reference/com/google/android/gms/location/FusedLocationProviderClient) is a part of Google's Location Services that uses Wi-Fi and GPS to *efficiently* obtain user location.

- The `FusedLocationProviderClient` [dependency is created](https://github.com/kyle-gonzales/jeepni/pull/31/commits/cfde16f5276cb31d1f8676851611f24ce3084d92#diff-303321dcf6421171359f5c70b83b2c89a0a2f9d02fe11c3d4b7e751bf200280eR68-R72) and is [injected](https://github.com/kyle-gonzales/jeepni/pull/31/commits/cfde16f5276cb31d1f8676851611f24ce3084d92#diff-303321dcf6421171359f5c70b83b2c89a0a2f9d02fe11c3d4b7e751bf200280eR68-R72) into the `MainViewModel`.
- When driver mode is toggled on, the user's current location is automatically requested ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/3967dea9b162cfe936b9debf5f04e8cb869e89b7#diff-dc3bad78ace96db9e9c983fbadf2d801fa350087c6bc36e11e02759d5b57dc55R127-R130))
  - The LocationProvider continues to collect updates on the user's location at set intervals ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/3967dea9b162cfe936b9debf5f04e8cb869e89b7#diff-dc3bad78ace96db9e9c983fbadf2d801fa350087c6bc36e11e02759d5b57dc55R127-R130))
  - Optimized performance of location provider by setting the minimum update distance to 10 meters to avoid receiving location updates when driver is not moving. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/2b51a622d3476bb7837d0fb3bd902533914e16da#diff-dc3bad78ace96db9e9c983fbadf2d801fa350087c6bc36e11e02759d5b57dc55R232))
- The marker on the map is updated and the map is animated every time a new location is obtained ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/3967dea9b162cfe936b9debf5f04e8cb869e89b7#diff-dc3bad78ace96db9e9c983fbadf2d801fa350087c6bc36e11e02759d5b57dc55R127-R130))
  - When the driver mode is initially toggled on, the initial camera position state of the map composable and marker composable are set to the coordinates of Cebu City as a *temporary*. This is because the Location Provider cannot provider the user's location in time for the map to load its initial state. However, this configuration can be improved in the future. ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/f2e4826f6d4936c4f3990009cf8a685f60c57143#diff-dc3bad78ace96db9e9c983fbadf2d801fa350087c6bc36e11e02759d5b57dc55R72-R77))

---

### Additional Changes (April 17, 2023)

- Driver mode is automatically toggled on when permissions are granted ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/d93fd907acf4dcd84908d8627f71520b3f348e76))
  - Permission requests are launched when the toggle button is pressed. However, the toggle button remains off even after permissions are granted. The switch has to be pressed again so that driver mode can finally be toggled on. In this commit, logic is added to the activity result callback. Now, granting permissions automatically turns driver mode on. We save one user press. lol. 
- Fixed the UI of the distance and time trackers on the `MainScreen` action bar ([changes](https://github.com/kyle-gonzales/jeepni/pull/31/commits/ce9a2a5b4a0db83f8f7654e04a8f6b51ea500b90))
  - Decreased the font size and changed the font family to be consistent with the rest of JeepNi

### CODE REVIEW

File: app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt [hunk 1]
```diff
@@ -40,12 +56,38 @@ fun MainScreen(
...
+    }
+    val multiplePermissionResultLauncher = rememberLauncherForActivityResult(
+        contract = ActivityResultContracts.RequestMultiplePermissions()) { permissions ->
+        var isAllGranted = false
+        permissions.keys.forEach { permission ->
+            viewModel.onPermissionResult(
+                permission = permission,
+                isGranted = permissions[permission] == true
+            )
+            isAllGranted = permissions[permission] == true
```
Comment (AmGthrow) [inline app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt]: Can you explain what's going on here @kyle-gonzales ? It looks like `isAllGranted` will only get the permission state of the last permission in the loop, not all permissions

File: app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt [hunk 2]
```diff
@@ -158,10 +213,39 @@ fun MainScreen(
                             viewModel.onEvent(MainEvent.OnSaveDailyAnalyticClick(salary.toDouble(), fuelCost.toDouble()))}
                     )
                 }
+                val activity = LocalContext.current as MainActivity //TODO: find another solution. this is probably a memory leak..
```
Comment (AmGthrow) [inline app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt]: I'm not sure how you can tell it's a memory leak pero if you're sure then we should tag it as an issue so it doesn't get forgotten. If you're not sure though we can pretend it doesn't exist until we actually start seriously profiling the app for leaks na

File: app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt (hunk 1 above)
Comment (kyle-gonzales) [reply app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt]: Oh my, thanks for catching that!. I'll fix the `isAllGranted` flag!

Comment (kyle-gonzales) [reply app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt]: Fixed in [this](https://github.com/kyle-gonzales/jeepni/pull/31/commits/0ec9829516e8de46262f58d2a41621c0ad9c7e1b) commit.

File: app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt (hunk 2 above)
Comment (kyle-gonzales) [reply app/src/main/java/com/example/jeepni/feature/home/MainScreen.kt]: Thanks for pointing this out! I actually added a [memory leak library](https://square.github.io/leakcanary/getting_started/) to our debug test and it doesn't seem to cry whenever I try to trigger the 'leak'. So I guess it's not a leak!

**Pull request merged**